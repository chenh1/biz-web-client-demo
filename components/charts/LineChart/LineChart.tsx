import type { EventType } from '@visx/event/lib/types';
import type { TouchEventHandler } from 'react';
import React, { useCallback } from 'react';
import { Group } from '@visx/group';
import { curveCardinal } from '@visx/curve';
import { LinePath, Line, Bar } from '@visx/shape';
import { Threshold } from '@visx/threshold';
import { withTooltip, Tooltip, TooltipWithBounds, defaultStyles } from '@visx/tooltip';
import { localPoint } from '@visx/event';
import { scaleTime, scaleLinear } from '@visx/scale';
import { AxisLeft, AxisBottom } from '@visx/axis';
import { GridRows, GridColumns } from '@visx/grid';
import { fontFamilyHeadings } from '../../constants/typography';
import { bisector } from '@visx/vendor/d3-array';

const defaultMargin = { top: 40, right: 30, bottom: 50, left: 40 };

export type LineChartProps = {
  width: number;
  height: number;
  margin?: typeof defaultMargin;
  tooltipMargin?: { top: number; right: number; bottom: number; left: number };
  data: any[];
  yAccessor: (d: any) => number;
  x1Accessor: (d: any) => number;
  x2Accessor?: (d: any) => number;
  yAxisLabel?: string;
  showTooltip?: (tooltip: { tooltipData: any; tooltipLeft: number; tooltipTop: number }) => void;
  hideTooltip?: () => void;
  tooltipData?: any;
  tooltipTop?: number;
  tooltipLeft?: number;
  paintThreshold?: boolean;
  thresholdOpacity?: number;
  shouldBeCurveLinear?: boolean;
  shouldSnapXToGrid?: boolean;
  aboveAreaFill?: string;
  belowAreaFill?: string;
  background?: string;
  generalStroke?: string;
  gridColor?: string;
  gridOpacity?: number;
  onClick?: (event: React.MouseEvent<SVGRectElement>, data: any) => void;
}

export const LineChart = withTooltip<LineChartProps>(
  ({
    width,
    height,
    margin = defaultMargin,
    tooltipMargin = { top: 0, right: 70, bottom: 90, left: 0 },
    data,
    yAccessor,
    x1Accessor,
    x2Accessor,
    yAxisLabel,
    showTooltip,
    hideTooltip,
    tooltipData,
    tooltipTop = 0,
    tooltipLeft = 0,
    paintThreshold,
    thresholdOpacity = 0.4,
    shouldBeCurveLinear,
    shouldSnapXToGrid = true,
    aboveAreaFill,
    belowAreaFill,
    background,
    generalStroke = '#222',
    gridColor = '#e0e0e0',
    gridOpacity = 1,
    onClick
  }) => {
    if (width < 10) return null;

    const tooltipStyles = {
      ...defaultStyles,
      background,
      border: `1px solid ${generalStroke}`,
      color: generalStroke,
      fontFamily: fontFamilyHeadings,
    };

    const bisectY = bisector(yAccessor).left;

    // Tooltip bounds
    const innerWidth = width - tooltipMargin.left - tooltipMargin.right;
    const innerHeight = height - tooltipMargin.top - tooltipMargin.bottom;

    // Scales
    const yScale = scaleTime<number>({
      domain: [Math.min(...data.map(yAccessor)), Math.max(...data.map(yAccessor))],
    });

    const xScale = scaleLinear<number>({
      domain: x2Accessor
        ? [
            Math.min(...data.map((d) => Math.min(x1Accessor(d), x2Accessor(d)))),
            Math.max(...data.map((d) => Math.max(x1Accessor(d), x2Accessor(d)))),
          ]
        : [
            Math.min(...data.map(x1Accessor)),
            Math.max(...data.map(x1Accessor)),
          ],
      nice: true,
    });

    // Bounds
    const xMax = width - margin.left - margin.right;
    const yMax = height - margin.top - margin.bottom;

    yScale.range([0, xMax]);
    xScale.range([yMax, 0]);

    // Tooltip handler
    const handleTooltip = useCallback(
      (event: Element | EventType) => {
        const { x: xRaw, y } = localPoint(event) || { x: 0, y: 0 };
        const x = xRaw - 40;
        const x0 = yScale.invert(x);

        const index = bisectY(data, x0, 1);
        const d0 = data[index - 1];
        const d1 = data[index];
        let d = d0;
        if (d1 && yAccessor(d1)) {
          d = x0.valueOf() - yAccessor(d0).valueOf() > yAccessor(d1).valueOf() - x0.valueOf() ? d1 : d0;
        }

        if (x2Accessor) {
          (d as any).x2TooltipTop = xScale(x2Accessor(d));
        }

        showTooltip({
          tooltipData: d,
          tooltipLeft: shouldSnapXToGrid ? yScale(yAccessor(d)) : x,
          tooltipTop: 20 + (xScale(x1Accessor(d)) * 0.95),
        });
      },
      [showTooltip, xScale, yScale, bisectY, data, x2Accessor, yAccessor, shouldSnapXToGrid]
    );

    return (
      <div>
        <svg width={width} height={height}>
          <rect x={0} y={0} width={width} height={height} fill={background} rx={14} />
          <Group left={margin.left} top={margin.top}>
            <GridRows scale={xScale} width={xMax} height={yMax} stroke={gridColor} opacity={gridOpacity} />
            <GridColumns scale={yScale} width={xMax} height={yMax} stroke={gridColor} opacity={gridOpacity} />
            <line x1={xMax} x2={xMax} y1={0} y2={yMax} stroke="#e0e0e0" />
            <AxisBottom
              stroke={generalStroke}
              tickLabelProps={{ fill: generalStroke, fontSize: 14, fontFamily: fontFamilyHeadings }}
              top={yMax}
              scale={yScale}
              numTicks={width > 520 ? 10 : 5}
            />
            <AxisLeft
              stroke={generalStroke}
              tickLabelProps={{ fill: generalStroke, fontSize: 14, fontFamily: fontFamilyHeadings }}
              scale={xScale}
            />
            <text x="-60" y="25" transform="rotate(-90)" fontSize={20} fontFamily={fontFamilyHeadings}>
              {yAxisLabel}
            </text>
            {paintThreshold && x2Accessor && (
              <Threshold
                id={`${Math.random()}`}
                data={data}
                x={(d) => yScale(yAccessor(d)) ?? 0}
                y0={(d) => xScale(x1Accessor(d)) ?? 0}
                y1={(d) => xScale(x2Accessor(d)) ?? 0}
                clipAboveTo={0}
                clipBelowTo={yMax}
                curve={shouldBeCurveLinear ? undefined : curveCardinal}
                belowAreaProps={{
                  fill: belowAreaFill,
                  fillOpacity: thresholdOpacity,
                }}
                aboveAreaProps={{
                  fill: aboveAreaFill,
                  fillOpacity: thresholdOpacity,
                }}
              />
            )}
            {x2Accessor && (
              <LinePath
                data={data}
                curve={shouldBeCurveLinear ? undefined : curveCardinal}
                x={(d) => yScale(yAccessor(d)) ?? 0}
                y={(d) => xScale(x2Accessor(d)) ?? 0}
                stroke={generalStroke}
                strokeWidth={1.5}
                strokeOpacity={0.8}
                strokeDasharray="1,2"
              />
            )}
            <LinePath
              data={data}
              curve={shouldBeCurveLinear ? curveCardinal : undefined}
              x={(d) => yScale(yAccessor(d)) ?? 0}
              y={(d) => xScale(x1Accessor(d)) ?? 0}
              stroke={generalStroke}
              strokeWidth={1.5}
            />
            <Bar
              x={tooltipMargin.left}
              y={tooltipMargin.top}
              width={innerWidth}
              height={innerHeight}
              fill="transparent"
              onTouchStart={handleTooltip}
              onTouchMove={handleTooltip}
              onMouseMove={handleTooltip}
              onMouseLeave={hideTooltip}
              onClick={(event) => onClick && onClick(event, tooltipData)}
            />
            {tooltipData && (
              <g>
                <Line
                  from={{ x: tooltipLeft, y: tooltipMargin.top }}
                  to={{ x: tooltipLeft, y: innerHeight + tooltipMargin.top }}
                  stroke={generalStroke}
                  strokeWidth={2}
                  pointerEvents="none"
                  strokeDasharray="5,2"
                />
                <circle
                  cx={tooltipLeft}
                  cy={tooltipTop + 1}
                  r={4}
                  fill="black"
                  fillOpacity={0.1}
                  stroke="black"
                  strokeOpacity={0.1}
                  strokeWidth={2}
                  pointerEvents="none"
                />
                <circle
                  cx={tooltipLeft}
                  cy={tooltipTop}
                  r={4}
                  fill={generalStroke}
                  stroke="white"
                  strokeWidth={2}
                  pointerEvents="none"
                />
                {tooltipData?.x2TooltipTop && (
                  <>
                    <circle
                      cx={tooltipLeft}
                      cy={tooltipData.x2TooltipTop + 1}
                      r={4}
                      fill="black"
                      fillOpacity={0.1}
                      stroke="black"
                      strokeOpacity={0.1}
                      strokeWidth={2}
                      pointerEvents="none"
                    />
                    <circle
                      cx={tooltipLeft}
                      cy={tooltipData.x2TooltipTop}
                      r={4}
                      fill={generalStroke}
                      stroke="white"
                      strokeWidth={2}
                      pointerEvents="none"
                    />
                  </>
                )}
              </g>
            )}
          </Group>
        </svg>
        {tooltipData && (
          <div>
            <TooltipWithBounds
              key={Math.random()}
              top={tooltipTop - 12}
              left={tooltipLeft + 12}
              style={tooltipStyles}
            >
              {`${x1Accessor(tooltipData)}`}
            </TooltipWithBounds>
            {tooltipData?.x2TooltipTop && x2Accessor && (
              <TooltipWithBounds
                key={Math.random()}
                top={tooltipData.x2TooltipTop - 12}
                left={tooltipLeft + 12}
                style={tooltipStyles}
              >
                {`${x2Accessor(tooltipData)}`}
              </TooltipWithBounds>
            )}
          </div>
        )}
      </div>
    );
  }
);
