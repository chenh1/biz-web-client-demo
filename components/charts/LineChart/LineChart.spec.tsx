import { LineChart } from "./LineChart";
import { render, waitFor } from '@testing-library/react';
import colors from '../../constants/colors'
import { mockData } from './LineChart.stories'
import { JSDOM } from "jsdom"
const dom = new JSDOM()
global.document = dom.window.document
global.window = dom.window


it('renders a Line chart given the dataset', async () => {
  const { asFragment } = render(
    <LineChart
      data={ mockData }
      yAccessor={(d) => new Date(d.date).valueOf()}
      x1Accessor={(d) => Number(d['CSCO'])}
      x2Accessor={(d) => Number(d['VWAP'])}
      yAxisLabel={'CSCO'}
      paintThreshold={true}
      shouldBeCurveLinear={false}
      aboveAreaFill={colors.cyberTeal}
      belowAreaFill={colors.machIndigo}
      background={colors.nightBlack}
      generalStroke={colors.hyperRed}
      gridColor={colors.hyperRed}
      gridOpacity={0.3}
      thresholdOpacity={0.8}
      width={500}
      height={300}
    />
  );

  await waitFor(() => {
    const frag = asFragment()
    const paths = [...frag.querySelectorAll('defs'), ...frag.querySelectorAll('path')];

    // remove non-deterministic elements (path and clippath's dynamic ids)
    paths.forEach((path) => {
      if (!!path && !!path.parentNode) {
        path.parentNode.removeChild(path);
      }
    });

    expect(frag).toMatchSnapshot();
  });
});