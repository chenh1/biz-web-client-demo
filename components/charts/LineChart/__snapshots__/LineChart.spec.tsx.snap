// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders a Line chart given the dataset 1`] = `
<DocumentFragment>
  <div
    style="position: relative; width: inherit; height: inherit;"
  >
    <div>
      <svg
        height="300"
        width="500"
      >
        <rect
          fill="#050E1C"
          height="300"
          rx="14"
          width="500"
          x="0"
          y="0"
        />
        <g
          class="visx-group"
          transform="translate(40, 40)"
        >
          <g
            class="visx-group visx-rows"
            transform="translate(0, 0)"
          >
            <line
              class="visx-line"
              fill="transparent"
              height="210"
              opacity="0.3"
              shape-rendering="crispEdges"
              stroke="#FF2F97"
              stroke-width="1"
              x1="0"
              x2="430"
              y1="210"
              y2="210"
            />
            <line
              class="visx-line"
              fill="transparent"
              height="210"
              opacity="0.3"
              shape-rendering="crispEdges"
              stroke="#FF2F97"
              stroke-width="1"
              x1="0"
              x2="430"
              y1="193.84615384615387"
              y2="193.84615384615387"
            />
            <line
              class="visx-line"
              fill="transparent"
              height="210"
              opacity="0.3"
              shape-rendering="crispEdges"
              stroke="#FF2F97"
              stroke-width="1"
              x1="0"
              x2="430"
              y1="177.69230769230768"
              y2="177.69230769230768"
            />
            <line
              class="visx-line"
              fill="transparent"
              height="210"
              opacity="0.3"
              shape-rendering="crispEdges"
              stroke="#FF2F97"
              stroke-width="1"
              x1="0"
              x2="430"
              y1="161.53846153846152"
              y2="161.53846153846152"
            />
            <line
              class="visx-line"
              fill="transparent"
              height="210"
              opacity="0.3"
              shape-rendering="crispEdges"
              stroke="#FF2F97"
              stroke-width="1"
              x1="0"
              x2="430"
              y1="145.3846153846154"
              y2="145.3846153846154"
            />
            <line
              class="visx-line"
              fill="transparent"
              height="210"
              opacity="0.3"
              shape-rendering="crispEdges"
              stroke="#FF2F97"
              stroke-width="1"
              x1="0"
              x2="430"
              y1="129.23076923076923"
              y2="129.23076923076923"
            />
            <line
              class="visx-line"
              fill="transparent"
              height="210"
              opacity="0.3"
              shape-rendering="crispEdges"
              stroke="#FF2F97"
              stroke-width="1"
              x1="0"
              x2="430"
              y1="113.07692307692307"
              y2="113.07692307692307"
            />
            <line
              class="visx-line"
              fill="transparent"
              height="210"
              opacity="0.3"
              shape-rendering="crispEdges"
              stroke="#FF2F97"
              stroke-width="1"
              x1="0"
              x2="430"
              y1="96.92307692307693"
              y2="96.92307692307693"
            />
            <line
              class="visx-line"
              fill="transparent"
              height="210"
              opacity="0.3"
              shape-rendering="crispEdges"
              stroke="#FF2F97"
              stroke-width="1"
              x1="0"
              x2="430"
              y1="80.76923076923076"
              y2="80.76923076923076"
            />
            <line
              class="visx-line"
              fill="transparent"
              height="210"
              opacity="0.3"
              shape-rendering="crispEdges"
              stroke="#FF2F97"
              stroke-width="1"
              x1="0"
              x2="430"
              y1="64.61538461538461"
              y2="64.61538461538461"
            />
            <line
              class="visx-line"
              fill="transparent"
              height="210"
              opacity="0.3"
              shape-rendering="crispEdges"
              stroke="#FF2F97"
              stroke-width="1"
              x1="0"
              x2="430"
              y1="48.46153846153845"
              y2="48.46153846153845"
            />
            <line
              class="visx-line"
              fill="transparent"
              height="210"
              opacity="0.3"
              shape-rendering="crispEdges"
              stroke="#FF2F97"
              stroke-width="1"
              x1="0"
              x2="430"
              y1="32.30769230769231"
              y2="32.30769230769231"
            />
            <line
              class="visx-line"
              fill="transparent"
              height="210"
              opacity="0.3"
              shape-rendering="crispEdges"
              stroke="#FF2F97"
              stroke-width="1"
              x1="0"
              x2="430"
              y1="16.153846153846143"
              y2="16.153846153846143"
            />
            <line
              class="visx-line"
              fill="transparent"
              height="210"
              opacity="0.3"
              shape-rendering="crispEdges"
              stroke="#FF2F97"
              stroke-width="1"
              x1="0"
              x2="430"
              y1="0"
              y2="0"
            />
          </g>
          <g
            class="visx-group visx-columns"
            transform="translate(0, 0)"
          >
            <line
              class="visx-line"
              fill="transparent"
              opacity="0.3"
              shape-rendering="crispEdges"
              stroke="#FF2F97"
              stroke-width="1"
              width="430"
              x1="32.125901875901874"
              x2="32.125901875901874"
              y1="0"
              y2="210"
            />
            <line
              class="visx-line"
              fill="transparent"
              opacity="0.3"
              shape-rendering="crispEdges"
              stroke="#FF2F97"
              stroke-width="1"
              width="430"
              x1="66.00468975468975"
              x2="66.00468975468975"
              y1="0"
              y2="210"
            />
            <line
              class="visx-line"
              fill="transparent"
              opacity="0.3"
              shape-rendering="crispEdges"
              stroke="#FF2F97"
              stroke-width="1"
              width="430"
              x1="100.255772005772"
              x2="100.255772005772"
              y1="0"
              y2="210"
            />
            <line
              class="visx-line"
              fill="transparent"
              opacity="0.3"
              shape-rendering="crispEdges"
              stroke="#FF2F97"
              stroke-width="1"
              width="430"
              x1="134.52236652236653"
              x2="134.52236652236653"
              y1="0"
              y2="210"
            />
            <line
              class="visx-line"
              fill="transparent"
              opacity="0.3"
              shape-rendering="crispEdges"
              stroke="#FF2F97"
              stroke-width="1"
              width="430"
              x1="168.01334776334775"
              x2="168.01334776334775"
              y1="0"
              y2="210"
            />
            <line
              class="visx-line"
              fill="transparent"
              opacity="0.3"
              shape-rendering="crispEdges"
              stroke="#FF2F97"
              stroke-width="1"
              width="430"
              x1="201.89213564213563"
              x2="201.89213564213563"
              y1="0"
              y2="210"
            />
            <line
              class="visx-line"
              fill="transparent"
              opacity="0.3"
              shape-rendering="crispEdges"
              stroke="#FF2F97"
              stroke-width="1"
              width="430"
              x1="236.1432178932179"
              x2="236.1432178932179"
              y1="0"
              y2="210"
            />
            <line
              class="visx-line"
              fill="transparent"
              opacity="0.3"
              shape-rendering="crispEdges"
              stroke="#FF2F97"
              stroke-width="1"
              width="430"
              x1="270.40981240981245"
              x2="270.40981240981245"
              y1="0"
              y2="210"
            />
            <line
              class="visx-line"
              fill="transparent"
              opacity="0.3"
              shape-rendering="crispEdges"
              stroke="#FF2F97"
              stroke-width="1"
              width="430"
              x1="303.9007936507937"
              x2="303.9007936507937"
              y1="0"
              y2="210"
            />
            <line
              class="visx-line"
              fill="transparent"
              opacity="0.3"
              shape-rendering="crispEdges"
              stroke="#FF2F97"
              stroke-width="1"
              width="430"
              x1="337.77958152958155"
              x2="337.77958152958155"
              y1="0"
              y2="210"
            />
            <line
              class="visx-line"
              fill="transparent"
              opacity="0.3"
              shape-rendering="crispEdges"
              stroke="#FF2F97"
              stroke-width="1"
              width="430"
              x1="372.03066378066376"
              x2="372.03066378066376"
              y1="0"
              y2="210"
            />
            <line
              class="visx-line"
              fill="transparent"
              opacity="0.3"
              shape-rendering="crispEdges"
              stroke="#FF2F97"
              stroke-width="1"
              width="430"
              x1="406.2972582972583"
              x2="406.2972582972583"
              y1="0"
              y2="210"
            />
          </g>
          <line
            stroke="#e0e0e0"
            x1="430"
            x2="430"
            y1="0"
            y2="210"
          />
          <g
            class="visx-group visx-axis visx-axis-bottom"
            transform="translate(0, 210)"
          >
            <g
              class="visx-group visx-axis-tick"
              transform="translate(0, 0)"
            >
              <line
                class="visx-line"
                fill="transparent"
                shape-rendering="crispEdges"
                stroke="#222"
                stroke-linecap="square"
                stroke-width="1"
                x1="134.52236652236653"
                x2="134.52236652236653"
                y1="0"
                y2="8"
              />
              <svg
                font-size="14"
                style="overflow: visible;"
                x="0"
                y="0.25em"
              >
                <text
                  fill="#FF2F97"
                  font-family=""Space Grotesk", sans-serif"
                  font-size="14"
                  text-anchor="middle"
                  transform=""
                  x="134.52236652236653"
                  y="22"
                >
                  <tspan
                    dy="0em"
                    x="134.52236652236653"
                  >
                    2018
                  </tspan>
                </text>
              </svg>
            </g>
            <g
              class="visx-group visx-axis-tick"
              transform="translate(0, 0)"
            >
              <line
                class="visx-line"
                fill="transparent"
                shape-rendering="crispEdges"
                stroke="#222"
                stroke-linecap="square"
                stroke-width="1"
                x1="270.40981240981245"
                x2="270.40981240981245"
                y1="0"
                y2="8"
              />
              <svg
                font-size="14"
                style="overflow: visible;"
                x="0"
                y="0.25em"
              >
                <text
                  fill="#FF2F97"
                  font-family=""Space Grotesk", sans-serif"
                  font-size="14"
                  text-anchor="middle"
                  transform=""
                  x="270.40981240981245"
                  y="22"
                >
                  <tspan
                    dy="0em"
                    x="270.40981240981245"
                  >
                    2019
                  </tspan>
                </text>
              </svg>
            </g>
            <g
              class="visx-group visx-axis-tick"
              transform="translate(0, 0)"
            >
              <line
                class="visx-line"
                fill="transparent"
                shape-rendering="crispEdges"
                stroke="#222"
                stroke-linecap="square"
                stroke-width="1"
                x1="406.2972582972583"
                x2="406.2972582972583"
                y1="0"
                y2="8"
              />
              <svg
                font-size="14"
                style="overflow: visible;"
                x="0"
                y="0.25em"
              >
                <text
                  fill="#FF2F97"
                  font-family=""Space Grotesk", sans-serif"
                  font-size="14"
                  text-anchor="middle"
                  transform=""
                  x="406.2972582972583"
                  y="22"
                >
                  <tspan
                    dy="0em"
                    x="406.2972582972583"
                  >
                    2020
                  </tspan>
                </text>
              </svg>
            </g>
            <line
              class="visx-line visx-axis-line"
              fill="transparent"
              shape-rendering="crispEdges"
              stroke="#FF2F97"
              stroke-width="1"
              x1="0.5"
              x2="430.5"
              y1="0"
              y2="0"
            />
          </g>
          <g
            class="visx-group visx-axis visx-axis-left"
            transform="translate(0, 0)"
          >
            <g
              class="visx-group visx-axis-tick"
              transform="translate(0, 0)"
            >
              <line
                class="visx-line"
                fill="transparent"
                shape-rendering="crispEdges"
                stroke="#222"
                stroke-linecap="square"
                stroke-width="1"
                x1="0"
                x2="-8"
                y1="210"
                y2="210"
              />
              <svg
                font-size="14"
                style="overflow: visible;"
                x="-0.25em"
                y="0.25em"
              >
                <text
                  fill="#FF2F97"
                  font-family=""Space Grotesk", sans-serif"
                  font-size="14"
                  text-anchor="end"
                  transform=""
                  x="-8"
                  y="210"
                >
                  <tspan
                    dy="0em"
                    x="-8"
                  >
                    52
                  </tspan>
                </text>
              </svg>
            </g>
            <g
              class="visx-group visx-axis-tick"
              transform="translate(0, 0)"
            >
              <line
                class="visx-line"
                fill="transparent"
                shape-rendering="crispEdges"
                stroke="#222"
                stroke-linecap="square"
                stroke-width="1"
                x1="0"
                x2="-8"
                y1="193.84615384615387"
                y2="193.84615384615387"
              />
              <svg
                font-size="14"
                style="overflow: visible;"
                x="-0.25em"
                y="0.25em"
              >
                <text
                  fill="#FF2F97"
                  font-family=""Space Grotesk", sans-serif"
                  font-size="14"
                  text-anchor="end"
                  transform=""
                  x="-8"
                  y="193.84615384615387"
                >
                  <tspan
                    dy="0em"
                    x="-8"
                  >
                    54
                  </tspan>
                </text>
              </svg>
            </g>
            <g
              class="visx-group visx-axis-tick"
              transform="translate(0, 0)"
            >
              <line
                class="visx-line"
                fill="transparent"
                shape-rendering="crispEdges"
                stroke="#222"
                stroke-linecap="square"
                stroke-width="1"
                x1="0"
                x2="-8"
                y1="177.69230769230768"
                y2="177.69230769230768"
              />
              <svg
                font-size="14"
                style="overflow: visible;"
                x="-0.25em"
                y="0.25em"
              >
                <text
                  fill="#FF2F97"
                  font-family=""Space Grotesk", sans-serif"
                  font-size="14"
                  text-anchor="end"
                  transform=""
                  x="-8"
                  y="177.69230769230768"
                >
                  <tspan
                    dy="0em"
                    x="-8"
                  >
                    56
                  </tspan>
                </text>
              </svg>
            </g>
            <g
              class="visx-group visx-axis-tick"
              transform="translate(0, 0)"
            >
              <line
                class="visx-line"
                fill="transparent"
                shape-rendering="crispEdges"
                stroke="#222"
                stroke-linecap="square"
                stroke-width="1"
                x1="0"
                x2="-8"
                y1="161.53846153846152"
                y2="161.53846153846152"
              />
              <svg
                font-size="14"
                style="overflow: visible;"
                x="-0.25em"
                y="0.25em"
              >
                <text
                  fill="#FF2F97"
                  font-family=""Space Grotesk", sans-serif"
                  font-size="14"
                  text-anchor="end"
                  transform=""
                  x="-8"
                  y="161.53846153846152"
                >
                  <tspan
                    dy="0em"
                    x="-8"
                  >
                    58
                  </tspan>
                </text>
              </svg>
            </g>
            <g
              class="visx-group visx-axis-tick"
              transform="translate(0, 0)"
            >
              <line
                class="visx-line"
                fill="transparent"
                shape-rendering="crispEdges"
                stroke="#222"
                stroke-linecap="square"
                stroke-width="1"
                x1="0"
                x2="-8"
                y1="145.3846153846154"
                y2="145.3846153846154"
              />
              <svg
                font-size="14"
                style="overflow: visible;"
                x="-0.25em"
                y="0.25em"
              >
                <text
                  fill="#FF2F97"
                  font-family=""Space Grotesk", sans-serif"
                  font-size="14"
                  text-anchor="end"
                  transform=""
                  x="-8"
                  y="145.3846153846154"
                >
                  <tspan
                    dy="0em"
                    x="-8"
                  >
                    60
                  </tspan>
                </text>
              </svg>
            </g>
            <g
              class="visx-group visx-axis-tick"
              transform="translate(0, 0)"
            >
              <line
                class="visx-line"
                fill="transparent"
                shape-rendering="crispEdges"
                stroke="#222"
                stroke-linecap="square"
                stroke-width="1"
                x1="0"
                x2="-8"
                y1="129.23076923076923"
                y2="129.23076923076923"
              />
              <svg
                font-size="14"
                style="overflow: visible;"
                x="-0.25em"
                y="0.25em"
              >
                <text
                  fill="#FF2F97"
                  font-family=""Space Grotesk", sans-serif"
                  font-size="14"
                  text-anchor="end"
                  transform=""
                  x="-8"
                  y="129.23076923076923"
                >
                  <tspan
                    dy="0em"
                    x="-8"
                  >
                    62
                  </tspan>
                </text>
              </svg>
            </g>
            <g
              class="visx-group visx-axis-tick"
              transform="translate(0, 0)"
            >
              <line
                class="visx-line"
                fill="transparent"
                shape-rendering="crispEdges"
                stroke="#222"
                stroke-linecap="square"
                stroke-width="1"
                x1="0"
                x2="-8"
                y1="113.07692307692307"
                y2="113.07692307692307"
              />
              <svg
                font-size="14"
                style="overflow: visible;"
                x="-0.25em"
                y="0.25em"
              >
                <text
                  fill="#FF2F97"
                  font-family=""Space Grotesk", sans-serif"
                  font-size="14"
                  text-anchor="end"
                  transform=""
                  x="-8"
                  y="113.07692307692307"
                >
                  <tspan
                    dy="0em"
                    x="-8"
                  >
                    64
                  </tspan>
                </text>
              </svg>
            </g>
            <g
              class="visx-group visx-axis-tick"
              transform="translate(0, 0)"
            >
              <line
                class="visx-line"
                fill="transparent"
                shape-rendering="crispEdges"
                stroke="#222"
                stroke-linecap="square"
                stroke-width="1"
                x1="0"
                x2="-8"
                y1="96.92307692307693"
                y2="96.92307692307693"
              />
              <svg
                font-size="14"
                style="overflow: visible;"
                x="-0.25em"
                y="0.25em"
              >
                <text
                  fill="#FF2F97"
                  font-family=""Space Grotesk", sans-serif"
                  font-size="14"
                  text-anchor="end"
                  transform=""
                  x="-8"
                  y="96.92307692307693"
                >
                  <tspan
                    dy="0em"
                    x="-8"
                  >
                    66
                  </tspan>
                </text>
              </svg>
            </g>
            <g
              class="visx-group visx-axis-tick"
              transform="translate(0, 0)"
            >
              <line
                class="visx-line"
                fill="transparent"
                shape-rendering="crispEdges"
                stroke="#222"
                stroke-linecap="square"
                stroke-width="1"
                x1="0"
                x2="-8"
                y1="80.76923076923076"
                y2="80.76923076923076"
              />
              <svg
                font-size="14"
                style="overflow: visible;"
                x="-0.25em"
                y="0.25em"
              >
                <text
                  fill="#FF2F97"
                  font-family=""Space Grotesk", sans-serif"
                  font-size="14"
                  text-anchor="end"
                  transform=""
                  x="-8"
                  y="80.76923076923076"
                >
                  <tspan
                    dy="0em"
                    x="-8"
                  >
                    68
                  </tspan>
                </text>
              </svg>
            </g>
            <g
              class="visx-group visx-axis-tick"
              transform="translate(0, 0)"
            >
              <line
                class="visx-line"
                fill="transparent"
                shape-rendering="crispEdges"
                stroke="#222"
                stroke-linecap="square"
                stroke-width="1"
                x1="0"
                x2="-8"
                y1="64.61538461538461"
                y2="64.61538461538461"
              />
              <svg
                font-size="14"
                style="overflow: visible;"
                x="-0.25em"
                y="0.25em"
              >
                <text
                  fill="#FF2F97"
                  font-family=""Space Grotesk", sans-serif"
                  font-size="14"
                  text-anchor="end"
                  transform=""
                  x="-8"
                  y="64.61538461538461"
                >
                  <tspan
                    dy="0em"
                    x="-8"
                  >
                    70
                  </tspan>
                </text>
              </svg>
            </g>
            <g
              class="visx-group visx-axis-tick"
              transform="translate(0, 0)"
            >
              <line
                class="visx-line"
                fill="transparent"
                shape-rendering="crispEdges"
                stroke="#222"
                stroke-linecap="square"
                stroke-width="1"
                x1="0"
                x2="-8"
                y1="48.46153846153845"
                y2="48.46153846153845"
              />
              <svg
                font-size="14"
                style="overflow: visible;"
                x="-0.25em"
                y="0.25em"
              >
                <text
                  fill="#FF2F97"
                  font-family=""Space Grotesk", sans-serif"
                  font-size="14"
                  text-anchor="end"
                  transform=""
                  x="-8"
                  y="48.46153846153845"
                >
                  <tspan
                    dy="0em"
                    x="-8"
                  >
                    72
                  </tspan>
                </text>
              </svg>
            </g>
            <g
              class="visx-group visx-axis-tick"
              transform="translate(0, 0)"
            >
              <line
                class="visx-line"
                fill="transparent"
                shape-rendering="crispEdges"
                stroke="#222"
                stroke-linecap="square"
                stroke-width="1"
                x1="0"
                x2="-8"
                y1="32.30769230769231"
                y2="32.30769230769231"
              />
              <svg
                font-size="14"
                style="overflow: visible;"
                x="-0.25em"
                y="0.25em"
              >
                <text
                  fill="#FF2F97"
                  font-family=""Space Grotesk", sans-serif"
                  font-size="14"
                  text-anchor="end"
                  transform=""
                  x="-8"
                  y="32.30769230769231"
                >
                  <tspan
                    dy="0em"
                    x="-8"
                  >
                    74
                  </tspan>
                </text>
              </svg>
            </g>
            <g
              class="visx-group visx-axis-tick"
              transform="translate(0, 0)"
            >
              <line
                class="visx-line"
                fill="transparent"
                shape-rendering="crispEdges"
                stroke="#222"
                stroke-linecap="square"
                stroke-width="1"
                x1="0"
                x2="-8"
                y1="16.153846153846143"
                y2="16.153846153846143"
              />
              <svg
                font-size="14"
                style="overflow: visible;"
                x="-0.25em"
                y="0.25em"
              >
                <text
                  fill="#FF2F97"
                  font-family=""Space Grotesk", sans-serif"
                  font-size="14"
                  text-anchor="end"
                  transform=""
                  x="-8"
                  y="16.153846153846143"
                >
                  <tspan
                    dy="0em"
                    x="-8"
                  >
                    76
                  </tspan>
                </text>
              </svg>
            </g>
            <g
              class="visx-group visx-axis-tick"
              transform="translate(0, 0)"
            >
              <line
                class="visx-line"
                fill="transparent"
                shape-rendering="crispEdges"
                stroke="#222"
                stroke-linecap="square"
                stroke-width="1"
                x1="0"
                x2="-8"
                y1="0"
                y2="0"
              />
              <svg
                font-size="14"
                style="overflow: visible;"
                x="-0.25em"
                y="0.25em"
              >
                <text
                  fill="#FF2F97"
                  font-family=""Space Grotesk", sans-serif"
                  font-size="14"
                  text-anchor="end"
                  transform=""
                  x="-8"
                  y="0"
                >
                  <tspan
                    dy="0em"
                    x="-8"
                  >
                    78
                  </tspan>
                </text>
              </svg>
            </g>
            <line
              class="visx-line visx-axis-line"
              fill="transparent"
              shape-rendering="crispEdges"
              stroke="#FF2F97"
              stroke-width="1"
              x1="0"
              x2="0"
              y1="210.5"
              y2="0.5"
            />
          </g>
          <text
            font-family=""Space Grotesk", sans-serif"
            font-size="20"
            transform="rotate(-90)"
            x="-60"
            y="25"
          >
            CSCO
          </text>
          <g
            class="visx-threshold"
          >
            <g />
          </g>
          <rect
            class="visx-bar"
            fill="transparent"
            height="210"
            width="430"
            x="0"
            y="0"
          />
        </g>
      </svg>
    </div>
  </div>
</DocumentFragment>
`;
